huaweicloud/services/vpn/resource_huaweicloud_vpn_user_group.go

```diff
@@ -0,0 +1,440 @@
+package vpn
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/chnsz/golangsdk"
+
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/common"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+)
+
+var userGroupNonUpdatableParams = []string{"vpn_server_id"}
+
+// @API VPN POST /v5/{project_id}/p2c-vpn-gateways/vpn-servers/{vpn_server_id}/groups
+// @API VPN GET /v5/{project_id}/p2c-vpn-gateways/vpn-servers/{vpn_server_id}/groups/{group_id}
+// @API VPN PUT /v5/{project_id}/p2c-vpn-gateways/vpn-servers/{vpn_server_id}/groups/{group_id}
+// @API VPN DELETE /v5/{project_id}/p2c-vpn-gateways/vpn-servers/{vpn_server_id}/groups/{group_id}
+// @API VPN POST /v5/{project_id}/p2c-vpn-gateways/vpn-servers/{vpn_server_id}/groups/{group_id}/add-users
+// @API VPN POST /v5/{project_id}/p2c-vpn-gateways/vpn-servers/{vpn_server_id}/groups/{group_id}/remove-users
+// @API VPN GET /v5/{project_id}/p2c-vpn-gateways/vpn-servers/{vpn_server_id}/groups/{group_id}/users
+func ResourceUserGroup() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceUserGroupCreate,
+		UpdateContext: resourceUserGroupUpdate,
+		ReadContext:   resourceUserGroupRead,
+		DeleteContext: resourceUserGroupDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceUserGroupImportState,
+		},
+
+		CustomizeDiff: config.FlexibleForceNew(userGroupNonUpdatableParams),
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"vpn_server_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `The VPN server ID.`,
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `The name of the user group.`,
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: `The description of the user group.`,
+			},
+			"users": {
```

Comment:
add a hash method? if the name or desc are updated, the users should not be updated